### DISCUSSION TITLE
Support of Mention/Suggestion extension

### DISCUSSION THREAD

**Discussion by leerobert** [sibiraj-s/svelte-tiptap#23] [General]
Similar to BubbleMenu, the Suggestion/Mention extension require injecting in a custom component. I played around with trying to get `SvelteRender` to work injecting a component in but I'm not smart enough with Svelte to figure this out.

**Comment by sibiraj-s** [comment]
Moved this to a discussion, as this is not an issue with the package

**Comment by sibiraj-s** [comment]
Based on what I see here https://tiptap.dev/api/nodes/mention

It would be really same instead of jsx and react renderer you would be using svelte component and svelte renderer, something like

Rendering the component
```js
import MentionList from './MentionList.svelte'

const mentionListComponent = new MentionList(props)

const renderer = SvelteRenderer(mentionListComponent, {element: yourDomHere })
```

To update props

```js
renderer.updateProps(props)
```

others are just tiptap stuff. I haven't tried this as I do not have enough time, but I believe the implementation is failrly similar and straightforward.

May be i will give it a shot sometime next week

**Reply by ysk3a** [reply]
If after giving a shot is successful, and when you have some time, would you be able to add a demo of it here? I would like see the implementation and also to try to use this and see how it will behave.

**Reply by leerobert** [reply]
https://github.com/sveltor/nextlint/blob/main/packages/svelte/src/lib/plugins/slash/suggestion.ts

**Reply by ysk3a** [reply]
i think that example would be good to follow if adding slash commands was needed but i was hoping for a more simpler example to start with that just shows list of users following https://tiptap.dev/api/nodes/mention example but for svelte. 
In the future i'll look into how to make it async to call from api but first I just wanted to see if tiptap suggestion in svelte would work.

more specifically, im having trouble understanding why 
```ts
const {clientRect, editor} = props;
      wrapper = document.createElement('div');
      editor.view.dom.parentNode?.appendChild(wrapper);
```
in https://github.com/sveltor/nextlint/blob/main/packages/svelte/src/lib/plugins/slash/renderer.ts
is needed and what it does.

**Comment by leerobert** [comment]
@sibiraj-s I found another project implementing suggestion into svelte:

https://github.com/sveltor/nextlint/blob/main/packages/svelte/src/lib/plugins/slash/suggestion.ts

Might be some inspiration here

**Reply by sibiraj-s** [reply]
Great. Thanks

**Comment by jcrd** [comment]
Hey, I am implementing the Mention extension using this discussion as a guideline. Nearly everything works as expected, but the query text used to filter suggestions remains part of the editor content rather than being replaced by the inserted Mention.

![image](https://github.com/sibiraj-s/svelte-tiptap/assets/44040597/087df32d-a284-45c6-9725-f3ad029c6d23)

I set up a [minimal reproduction](https://stackblitz.com/edit/svelte-tiptap-mention?file=src%2Flib%2Fsuggestion.js).

Any thoughts on what's going wrong?

**Reply by sibiraj-s** [reply]
Hi, you should be asking this over the tiptap channels. This is just the wrapper for making tiptap work with svelte.

Sorry for that.

**Reply by epulla** [reply]
hey @jcrd 👋 your code works perfectly! Here's a trick I used to remove the query used to filter the suggestions:

```javascript
onExit: (props) => {
          const range = {
		from: props.editor.state.selection.from,
		to: props.editor.state.selection.from + props.query.length
	 }
	 props.editor.commands.setTextSelection(range);
	 props.editor.commands.deleteSelection();
         popup[0].destroy();
         renderer.destroy();
         wrapper.remove();
},
```

Just replace your `onExit()` from your `suggestion.js` file with the code above. Here's a [forked project](https://stackblitz.com/edit/svelte-tiptap-mention-xhymu2?file=src%2Flib%2Fsuggestion.js) using your minimal reproduction.

Hope it helps 🙏

**Reply by sidharthv96** [reply]
@epulla I was a bit confused initially as I felt the key events in the list were not working. But it was actually working perfectly! I've [forked here](https://stackblitz.com/edit/svelte-tiptap-mention-m1pgyx?file=src%2Flib%2Flist.svelte) and added a bit of CSS to highlight the selection, and make it clickable as well.

**Reply by DanConwayDev** [reply]
Any suggestions for svelte 5? `new List()` is not valid.

**Reply by Wombosvideo** [reply]
@DanConwayDev 

If you are using TypeScript too, this should get you started quickly:

```typescript
// file: suggestion.svelte.ts <-- notice the rename, to add support for $state()

import { mount, unmount, type ComponentProps } from 'svelte';
import { type SuggestionOptions } from '@tiptap/suggestion';

// ...

export const suggestion: () => Omit<
  SuggestionOptions<
    string, // <-- Item type
    { // <-- Callback args / Command props
      id: string | null;
    }
  >,
  'editor'
> = () => ({
  // ...
  render: () => {
    let wrapper: HTMLDivElement;
    let componentProps: ComponentProps<typeof List> = $state(null!); // <-- initialize deeply reactive state
    let component: List;

    return {
      onStart: (props) => {

        // ...

        componentProps = { // <-- set initial component property values
          items: props.items,
          callback: (item) => {
            props.command({ id: item });
          },
        };

        component = mount(List, { // <-- create and mount component -- no need for SvelteRenderer
          target: wrapper,
          props: componentProps,
        });

        // ...
      },
      onUpdate: (props) => {
        componentProps.items = props.items; // <-- update the props directly

        // ...
      },
      // ...
      onExit: (props) => {
        // ...

        unmount(component); // <-- unmount after use
        wrapper.remove();
      },
    };
  },
});

export default suggestion;

**Comment by axel7083** [comment]
Hey, was wondering if there are some example combining [@tiptap/extension-mention](https://www.npmjs.com/package/@tiptap/extension-mention) **v3** with svelte-tiptap


I tried @jcrd but this requires tippy which has been removed in v3

**Reply by sibiraj-s** [reply]
hi @axel7083. No currently there are no examples. You should be able to find how floating menu UI is used with mentions plugin here https://tiptap.dev/docs/editor/extensions/nodes/mention#page-title and may be combine that with current example and get it working.

If you are able to get it working, do share here so it i'll help everyone.

**Reply by axel7083** [reply]
Hey, here is a demo https://axel7083.github.io/svelte-tiptap-mention/ (source code in https://github.com/axel7083/svelte-tiptap-mention)

This is not using `svelte-tiptap` I was able to make something works with _just_   `@floating-ui/dom`, `@tiptap/core` and `@tiptap/extension-mention`

this is not perfect, but will be enough for my need
